# **Cars 4 You | Machine Learning Project**

### **Import Libraries**

In [77]:
!pip install category_encoders
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import OrdinalEncoder
from sklearn.impute import KNNImputer
from rapidfuzz import process, fuzz
from category_encoders import TargetEncoder
from sklearn.model_selection import train_test_split
from datetime import datetime
from sklearn.preprocessing import StandardScaler


## **Normalization**

In [76]:
# Assuming your encoded dataset is called:
df_encoded = df_encoded.copy()

# numeric columns to normalize (excluding target "price")
num_cols_to_normalize = ['age', 'mileage', 'engine_size','tax', 'mpg','paint_quality']

# 'previousOwners' - count variable with small range

# Initialize the scaler
scaler = StandardScaler()

# Temporary normalization (for inspection only)
df_encoded[num_cols_to_normalize] = scaler.fit_transform(df_encoded[num_cols_to_normalize])
# Check results
df_encoded[num_cols_to_normalize].head(3)

age   mileage  engine_size       tax       mpg  paint_quality
0  0.507244  0.232346     0.607837 -0.297168 -2.807257      -0.070668
1 -0.869999 -0.875453    -0.280939  0.387221 -0.473669      -0.702760
2 -0.869999 -0.920310    -0.280939  0.387221 -0.921452      -0.411025

In [74]:
df_encoded[['age', 'mileage', 'engine_size','tax', 'mpg','paint_quality']].describe().round(2).T

count  mean  std   min   25%   50%   75%    max
age            75940.0  -0.0  1.0 -3.17 -0.87  0.05  0.51   9.69
mileage        75940.0  -0.0  1.0 -1.09 -0.73 -0.27  0.43  13.93
engine_size    75940.0   0.0  1.0 -2.95 -0.81 -0.10  0.61   6.83
tax            75940.0   0.0  1.0 -1.94  0.02  0.39  0.39   7.36
mpg            75940.0  -0.0  1.0 -3.47 -0.52 -0.02  0.48  26.58
paint_quality  75940.0   0.0  1.0 -3.04 -0.85  0.03  0.85   1.73

After applying StandardScaler all continuous features were centered (mean ≈ 0) and scaled (std ≈ 1) to ensure that all features contribute equally in the Lasso model.

- age: ranges roughly from -3.16 to 9.68. Some cars are much newer than the average or much older.
- mileage: ranges from -1.09 to 13.93. A few vehicles have extremely high mileage.
- engine_size: roughly -2.95 to 6.83. Some cars have much larger engines than the average.
- tax: roughly -1.94 to 7.36. Most cars are near average, but some high-tax vehicles stand out.
- mpg: roughly -3.46 to 2.65. Some vehicles are very fuel-inefficient (low MPG) or very efficient.
- paint_quality: -3.03 to 1.72. Some cars have very poor paint quality compared to the average.